### DISCUSSION TITLE
[no-misused-signals] restrict incorrect usage of signals - proposal

### DISCUSSION THREAD

**Discussion by MillerSvt** [angular-eslint/angular-eslint#2302] [New Rule Proposals]
**Description of the proposal**

This rule forbids providing Signals to logical locations such as if statements in places where the TypeScript compiler allows them but they are not handled properly. It could be about using signals in a nullish check or comparing them using operators like `===`, `==`, `>`, `<`, `>=`, or `<=`. These situations can often arise due to a missing call or just a misunderstanding of the way signals are handled.

In some rare cases `===` or `==` operators can be valid. So need to add option `ignoreNullishCheck` to disable higlighting them.

Reference: https://typescript-eslint.io/rules/no-misused-promises/

**Incorrect code**
```typescript
const signal = signal(0);
const otherSignal = signal(0);

if (signal) {
  // Do something
}

if (signal === otherSignal) {
  // Do something
}

if (signal > otherSignal) {
  // Do something
}

if (signal < otherSignal) {
  // Do something
}

const val = signal ? 123 : 456;

while (signal) {
  // Do something
}
```
```html
@if (signal) {
   something  
}

@if (signal === otherSignal) {
   something  
}

@if (signal > otherSignal) {
   something  
}

@if (signal < otherSignal) {
   something  
}

@switch (signal) {
   something
}

{{ signal }}
```

**Correct code**
```typescript
const signal = signal(0);
const otherSignal = signal(0);

if (signal()) {
  // Do something
}

if (signal() === otherSignal()) {
  // Do something
}

if (signal() > otherSignal()) {
  // Do something
}

if (signal() < otherSignal()) {
  // Do something
}

const val = signal() ? 123 : 456;

while (signal()) {
  // Do something
}
```
```html
@if (signal()) {
   something  
}

@if (signal() === otherSignal()) {
   something  
}

@if (signal() > otherSignal()) {
   something  
}

@if (signal() < otherSignal()) {
   something  
}

@switch (signal()) {
   something
}

{{ signal() }}
```

**Comment by PowerKiKi** [comment]
This seems to be a duplicate of #1753 and should probably be merged into a single issue.

**Comment by marcorohn** [comment]
Since signals are functions, this is also valid typescript code, but will never run, as `Function.length` always returns `0` for signals:

```typescript
const mySignal = signal(['hello', 'world']);
if (mySignal.length > 0) {
  ...
}
```

Just happened to me recently when converting a component to use signals, and I'm recently sure that is not that uncommon

**Comment by JamesHenry** [comment]
Thanks folks, it makes sense to me and I'll happily accept a PR for this rule

**Comment by pavel-ps** [comment]
Hi, I’d like to ask whether the fix has been merged and in which version. I haven’t been able to find it anywhere. Thank you in advance.

**Comment by rdlabo** [comment]
I released my own set of rules today, and I believe they could be useful for this proposal. It already covers quite a few of the ideas mentioned here. I'd be very grateful if you could check it out!

https://github.com/rdlabo-team/eslint-plugin-rules/releases/tag/v20.0.0

**Reply by rdlabo** [reply]
I’ve implemented a custom ESLint rule to check for Signals usage in templates, but I have to admit the approach isn’t ideal. When using templateUrl, the only way I found to validate the templates was to resolve and read the files manually using fs, since the template alone doesn’t provide enough context to determine whether something is a Signal.

As a result, linting becomes quite heavy. Unless there’s a better way to handle this, I’m starting to think that supporting templates might be better suited for third-party tools.

**Comment by SchroederSteffen** [comment]
~Please mark this as completed by #2383: https://github.com/angular-eslint/angular-eslint/blob/main/packages/eslint-plugin/docs/rules/no-uncalled-signals.md~

// Edit: This is partly completed by #2383 but misses templates: https://github.com/angular-eslint/angular-eslint/blob/main/packages/eslint-plugin/docs/rules/no-uncalled-signals.md

**Reply by MillerSvt** [reply]
Is this rule works in templates too?

**Reply by SchroederSteffen** [reply]
No. https://github.com/angular-eslint/angular-eslint/pull/2383#issuecomment-2935139383

**Reply by MillerSvt** [reply]
So, this issue isn't completed yet.

**Reply by SchroederSteffen** [reply]
You're right, I adjusted my response.

**Reply by SchroederSteffen** [reply]
See also #2647.

**Reply by venkatesh-nagineni** [reply]
this rule will not apply signal based queries like viewchild right

**Reply by PowerKiKi** [reply]
Correct, the rule `@angular-eslint/no-uncalled-signals` will **not** trigger when using a `viewChild` in a `if` condition (for example). I think it is a bug/missing feature, and angular-eslint should be fixed.

**Reply by MillerSvt** [reply]
Maybe we should open a separate issue.

**Reply by venkatesh-nagineni** [reply]
I will create a new issue then

**Reply by venkatesh-nagineni** [reply]
just create new issue https://github.com/angular-eslint/angular-eslint/issues/2904

**Comment by MillerSvt** [comment]
Closing this, as the behavior is already handled — in code by the [no-uncalled-signals](https://github.com/angular-eslint/angular-eslint/blob/main/packages/eslint-plugin/docs/rules/no-uncalled-signals.md) rule, and in templates through [extended diagnostics](https://angular.dev/extended-diagnostics/NG8109).
